### DISCUSSION TITLE
Is the display syntax tree format based on the existing specification or the custom specification of tree-sitter?

### DISCUSSION THREAD

**Discussion by liaodalin19903** [tree-sitter/tree-sitter#3477] [Q&A] [answered]
When I using playground:
https://tree-sitter.github.io/tree-sitter/playground

![b5cef501cb47b983d4f7d2647c8b7c3c](https://github.com/user-attachments/assets/e138be4a-6230-4012-871c-db4d68ac3e99)

Here is generating a syntax tree, I have 2 questions here:
1. Is this tree CST or AST?
2. Is the format here based on the existing specification or the custom specification of tree-sitter?
Such as: here are the keywords: program, variable_declaration, variable_declarator, identifier, and so on.
If it is a tree-sitter custom specification, where can I find detailed instructions?

**Comment by sogaiu** [comment]
Based on [this section](https://tree-sitter.github.io/tree-sitter/using-parsers#named-vs-anonymous-nodes) of the documentation, it seems likely that it's a CST:

> Tree-sitter produces [concrete syntax trees](https://en.wikipedia.org/wiki/Parse_tree) - trees that contain nodes for every individual token in the source code, including things like commas and parentheses.

AFAIU, for:

>  program, variable_declaration, variable_declarator, identifier, and so on

these things are specified in the tree-sitter grammar for JavaScript (based on your screenshot it looks that way anyway), for example, `program` is specified on [this line](https://github.com/tree-sitter/tree-sitter-javascript/blob/d767b1a276a4e80d7a6be30bada3070740ba4fa2/grammar.js#L109).

There is a [Creating parsers](https://tree-sitter.github.io/tree-sitter/creating-parsers) section of the official documentation that touches on some of these details.  May be you'll find that to be of some use.

**Reply by liaodalin19903** [reply]
thanks for your reply, when I switch the language to Python:
![cf8d27a1d7862b90f2adcd3cc6c7ec1b](https://github.com/user-attachments/assets/99d7f5ad-68f3-4a39-b0cf-238fc7a8ded1)

there comes different key words: `module, assignment ...` 
1、in the [py-tree-sitter](https://github.com/tree-sitter/py-tree-sitter), there is no `grammer.js` -like file to state the key workds. 

2、is there any `key words` document of the `syntax tree`?

**Reply by clason** [reply]
* `py-tree-sitter` is Python code for using tree-sitter (from Python)
* `tree-sitter-python` is the grammar for the Python language

and similar for other languages

And the node names are just what the grammar author chose; there's no "universal" list, and every grammar is different.

**Reply by liaodalin19903** [reply]
I find the `tree-sitter-python` has no grammar file like `tree-sitter-javascript`, 
but Python official document has the grammar definition: 
https://docs.python.org/3.10/reference/grammar.html

so the node names is defined by the `tree-sitter Parsers` author or the language author?

**Reply by sogaiu** [reply, marked as answer]
I see a `grammar.js` for tree-sitter-python [here](https://github.com/tree-sitter/tree-sitter-python/blob/0dee05ef958ba2eae88d1e65f24b33cad70d4367/grammar.js).

AFAIU, the author(s) of the grammar decides on what goes in `grammar.js` including names of nodes.  When there is an official language specification, I think the names contained within can influence choices the grammar author(s) make.

For languages that either have no formal specification or the specification is less "complete", new names may need to be created.

There is a [Creating parsers](https://tree-sitter.github.io/tree-sitter/creating-parsers) section linked earlier that has details regarding a possible grammar creation process.  I think going through it might give some related answers.
